en/docs/deploy-and-publish/deploy-on-gateway/choreo-connect/security/api-authorization/opa-validation.md

```diff
@@ -0,0 +1,203 @@
+
+```json tab='Format'
+{
+    "result": <BOOLEAN_POLICY_RESULT>
+}
+```
+
+```json tab='Sample'
+{
+    "result": true
+}
+```
+
+## Custom OPA Request Generator
+
+1.  Create a Java project with `org.wso2.choreo.connect.enforcer.commons` dependency.
+    -   For Apache Maven, use the following.
+        ```xml
+        <dependency>
+            <groupId>org.wso2.choreo.connect</groupId>
+            <artifactId>org.wso2.choreo.connect.enforcer.commons</artifactId>
+            <version>1.1.0</version>
+        </dependency>
+        ```
+
+2.  Use the following interface to implement the Custom OPA Request Generator.
+
+    ```java tab='Interface'
+    public interface OPARequestGenerator {}
+    ```
+
+    Here is the sample filter implementation that only returns request headers to the OPA.
+
+    ```java tab='Sample Implantation'
+    package org.example.tests;
+
+    import org.json.JSONException;
+    import org.json.JSONObject;
+    import org.wso2.choreo.connect.enforcer.commons.model.RequestContext;
+    import org.wso2.choreo.connect.enforcer.commons.opa.OPARequestGenerator;
+    import org.wso2.choreo.connect.enforcer.commons.opa.OPASecurityException;
+
+    public class CustomOPARequestGenerator implements OPARequestGenerator {
+        @Override
+        public String generateRequest(String policyName, String rule, Map<String, Object> advancedProperties,
+                                  RequestContext requestContext) throws OPASecurityException {
+            JSONObject requestPayload = new JSONObject();
+            JSONObject inputPayload = new JSONObject();
+            requestPayload.put("input", inputPayload);
+
+            inputPayload.put("headers", requestContext.getHeaders());
+            return requestPayload.toString();
+        }
+
+        @Override
+        public boolean validateResponse(String policyName, String rule, String opaResponse, RequestContext requestContext)
+                throws OPASecurityException {
+            JSONObject response = new JSONObject(opaResponse);
+            try {
+                return response.getBoolean("result");
+            } catch (JSONException e) {
+                throw new OPASecurityException(500, "Internal Server Error",
+                        "Error while evaluating remote authorization response", e);
+            }
+        }
+    }
+    ```
+
+3.  Since we use Java SPI (Service Provider Interface), we need to provide the provider configuration file `META-INF/services/org.wso2.choreo.connect.enforcer.commons.opa.OPARequestGenerator`. If you are using Apache Maven, create the file inside the `<PROJECT>/src/main/resources` directory. The content of the file needs to be the qualified class name of the filter implementation. For example:
```

Comment:
```suggestion
3.  Provide the provider configuration.

     As you use a Java SPI (Service Provider Interface), you need to provide the provider configuration file `META-INF/services/org.wso2.choreo.connect.enforcer.commons.opa.OPARequestGenerator`. 
     
     If you are using Apache Maven, create the file inside the `<PROJECT>/src/main/resources` directory. Add the qualified class name of the filter implementation as the content of the file. For example:
```

Reply:
@Mariangela Could you check the following? I changed the suggestion a little bit.

```suggestion
3.  Provide the OPARequestGenerator provider configuration.

    Choreo Connect uses Java SPI (Service Provider Interface) for loading the request generator implementation. Hence, you need to provide the provider configuration file `META-INF/services/org.wso2.choreo.connect.enforcer.commons.opa.OPARequestGenerator`. 
     
    If you are using Apache Maven, create the file inside the `<PROJECT>/src/main/resources` directory. Add the qualified class name of the OPARequestGenerator implementation as the content of the file. For example:
```